### ISSUE TITLE
Support bash autocomplete

### ISSUE THREAD

**Issue by rickypai** [philwo/bazelisk#29]
Since bazelisk downloads entire binaries, autocomplete scripts are not accessible. bazelisk could potentially also download the autocomplete scripts to a predetermined location to make it easier invoke.

**Comment by jin** [comment]
+1, this will be excellent to have.

**Comment by jin** [comment]
I suppose you can do this manually with instructions from https://docs.bazel.build/versions/master/completion.html#bash

```
$ bazel build //scripts:bazel-complete.bash
$ cp bazel-bin/scripts/bazel-complete.bash /etc/bash_completion.d/bazelisk

# new shell
$ bazelisk <TAB>
analyze                                                 --nodeep_execroot 
analyze-profile                                         --noexoblaze 
aquery                                                  --noexpand_configs_in_place 
--batch                                                 --noexperimental_oom_more_eagerly 
--batch_cpu_scheduling                                  --noidle_server_tasks 
--blazerc=                                              --noignore_all_rc_files 
..
```

**Comment by rickypai** [comment]
yeah that's what i ended up doing... I was hoping something like a bundled bash completion script that `bazelisk` could provide, but it's not super blocking

**Comment by philwo** [comment]
I would love to have this myself, but currently don't have time to implement this - if someone wants to contribute a PR, I would be very happy to review it.

**Comment by sudoforge** [comment]
I'll take a stab at this when I have time over the next few days.

**Comment by philwo** [comment]
@sudoforge Thanks! If you could find a solution that also works with zsh completions, I'd be extra grateful 😊 (I was a long-time bash user, but since macOS Catalina switched to zsh by default, I gave it a try and stuck with it.)

**Comment by sudoforge** [comment]
I'll be looking to add completion for all shells that Bazel supports -- off the top of my head, I think that's just Bash and ZSH.

**Comment by finn-ball** [comment]
+1 would be super useful to have

**Comment by gibfahn** [comment]
FWIW this is done now for zsh on Homebrew installs of bazelisk: https://github.com/bazelbuild/homebrew-tap/pull/89

**Comment by UebelAndre** [comment]
I've got it working on my mac using zsh but it's painfully slow. Are others experiencing the same?

**Comment by UebelAndre** [comment]
Having done some comparisons between bazel and bazelisk using zsh completion on my mac, bazelisk is orders of magnitude slower.

**Comment by UebelAndre** [comment]
> Having done some comparisons between bazel and bazelisk using zsh completion on my mac, bazelisk is orders of magnitude slower.

I wonder if https://github.com/bazelbuild/bazelisk/pull/248 would improve this

**Comment by sudoforge** [comment]
I've given some thought to this and will be sending a PR out in the next few days with a proposed solution.

**Comment by rpwoodbu** [comment]
Any update here? I'd be glad to help if I can.

**Comment by andrewring** [comment]
@sudoforge Is this still something you're planning to open?

**Comment by kylepl** [comment]
I was looking at this briefly, and my assumption as to what needs to be done:
* Update [DownloadRelease](https://github.com/bazelbuild/bazelisk/blob/24935f4ce663d6eb79e428005b763c059b508866/repositories/gcs.go#L141) to also fetch the auto-complete scripts
* To get the auto-complete scripts, the source needs to be downloaded (e.g. https://releases.bazel.build/6.0.0/release/bazel_6.0.0.tar.gz), and for simplicity probably just extract the entire "scripts/" directory (which includes bazel-complete-header.bash and such, and other auto-completes).
* Make that scripts/ sub-directory available in a deterministic place, so switching bazel versions also switches the underlying scripts

Seem reasonable to others?

**Comment by mmonaco** [comment]
Simply downloading the source and concatenating the header and template isn't exactly equivalent to building the completions properly. But maybe it's good enough?

Another way to look at this is bazel should have a (hidden?) `bazel completions <bash|fish|zsh>` subcommand like many other tools. That would make it easier to distribute the completions that are built properly. It would then still be nice if bazelisk had some support for using that output to keep a local file up to date, but wouldn't be as necessary because we'd all have a cleaner way to access the completions.
